Initial commits -- basic Helm chart for app of apps

sledigabel opened the pull request:
This commit gets us started.
The helm chart is generating application objects from a map of cluster
names and their corresponding endpoints.

There are only two options at the moment:
- `sequential`: applications will be rolled out one cluster after the
others.
- `parallel`: applications will be rolled out across clusters at the same
time.

We use sync wave to orchestrate the deployments.

This PR also embeds a limited testing + a github action configuration that would run lining.
Again this is simply to get us going.

sledigabel committed:
Inital commit

This commit gets us started.
The helm chart is generating application objects from a map of cluster
names and their corresponding endpoints.

There are only two options at the moment:
- sequential: applications will be rolled out one cluster after the
others.
- parallel: applications will be rolled out across clusters at the same
time.

We use sync wave to orchestrate the deployments.

sledigabel committed:
Adding initial CI and tests

sledigabel committed:
Removing default chart name, repo, revision values

Mistakenly left those values from testing.
We don't want a default chart, this needs to be from a user input.

danmx commented on .github/workflow/pullrequest.yaml:
```diff
@@ -0,0 +1,24 @@
+  ci:
+    runs-on: ubuntu-latest
+    env:
+      HELM_VERSION: "v3.3.2"
```
New version of Helm was recently released: `3.3.3`

sledigabel replied:
they _just_ released helm-3.3.4 :-)

danmx commented on .github/workflow/pullrequest.yaml:
```diff
@@ -0,0 +1,24 @@
+      uses: actions/checkout@v1
+    - name: Helm Install
+      run: |
+        curl -sSL https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz \
```
verify `tar`s hash:
```console
$ shasum -a 256 helm-v3.3.3-linux-amd64.tar.gz 
246d58b6b353e63ae8627415a7340089015e3eb542ff7b5ce124b0b1409369cc  helm-v3.3.3-linux-amd64.tar.gz
```

danmx commented on templates/application.yaml:
```diff
@@ -0,0 +1,35 @@
+      values: {{ nindent 7 (toYaml $.Values.remoteChart.values) }}
+    {{- end }}
+  {{- if $.Values.autosync }}
+  syncPolicy:
```
should it be configurable?

sledigabel replied:
good point.

danmx commented on templates/application.yaml:
```diff
@@ -0,0 +1,35 @@
+spec:
+  destination:
+    server: {{ $cluster_address | quote }}
+    namespace: default
```
should it be configurable or `{{ $.Release.namespace }}`?

sledigabel replied:
no; it's the target namespace. i'll add a config item.

danmx reviewed (changes requested):
Add chart README. You can autogenerate it using different tools, e.g.:
- https://github.com/norwoodj/helm-docs
- https://github.com/kubepack/chart-doc-gen

sledigabel committed:
Updating chart based on comments

- Added sha check after download
- updated helm version
- added target namespace, project and sync options

sledigabel commented:
@danmx are you happy to review again?

danmx commented on .github/workflow/pullrequest.yaml:
```diff
@@ -0,0 +1,26 @@
+    - name: Helm Install
+      run: |
+        curl -sSL https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz -o helm-${HELM_VERSION}-linux-amd64.tar.gz \
+        && curl -sSL https://get.helm.sh/helm-${HELM_VERSION}-linux-amd64.tar.gz.sha256sum -o helm-${HELM_VERSION}-linux-amd64.tar.gz.sha256sum \
```
the point is to hardcore the hash value abd just update it whenever we update the version

sledigabel committed:
Hardcoding the checksum

sledigabel commented:
got what you meant now, @danmx.
